pkg/apiserver/watch.go

```diff
@@ -64,119 +59,218 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
 	return t.C, t.Stop
 }
 
+type textEncodable interface {
+	// EncodesAsText should return true if objects should be transmitted as a WebSocket Text
+	// frame (otherwise, they will be sent as a Binary frame).
+	EncodesAsText() bool
+}
+
 // serveWatch handles serving requests to the server
+// TODO: the functionality in this method and in WatchServer.Serve is not cleanly decoupled.
 func serveWatch(watcher watch.Interface, scope RequestScope, req *restful.Request, res *restful.Response, timeout time.Duration) {
-	s, mediaType, err := negotiateOutputSerializer(req.Request, scope.Serializer)
+	// negotiate for the stream serializer
+	serializer, mediaType, err := negotiateOutputSerializer(req.Request, scope.StreamSerializer)
 	if err != nil {
 		scope.err(err, res.ResponseWriter, req.Request)
 		return
 	}
-	// TODO: replace with typed serialization
-	if mediaType != "application/json" {
-		writeRawJSON(http.StatusNotAcceptable, (errNotAcceptable{[]string{"application/json"}}).Status(), res.ResponseWriter)
+	encoder := scope.StreamSerializer.EncoderForVersion(serializer, scope.Kind.GroupVersion())
+
+	useTextFraming := false
+	if encodable, ok := encoder.(textEncodable); ok && encodable.EncodesAsText() {
+		useTextFraming = true
+	}
+
+	// find the embedded serializer matching the media type
+	// TODO: store content type on runtime.Unknown and serialize that into the stream, allowing for mixed payloads
```

Comment:
The protobuf serializer probably needs to be able to encode anything it can't recognize as a "nested" type (or just arbitrarily by being passed a parent JSON serializer).  Until we sort through what that looks like (which isn't implemented in this PR, but we do have a TODO there), it's possible that the embeddedSerializer might need to decide what to do with fallback cases.